mu/modes/microbit.py

```diff
@@ -226,20 +230,28 @@ def flash(self):
         """
         Takes the currently active tab, compiles the Python script therein into
         a hex file and flashes it all onto the connected device.
+
+        WARNING: This method is getting more complex due to several edge
+        cases. Ergo, it's a target for refactoring.
         """
-        logger.info('Flashing script.')
+        self.python_script = ''
+        logger.info('Preparing to flash script.')
+        # The first thing to do is check the script is valid and of the
+        # expected length.
         # Grab the Python script.
         tab = self.view.current_tab
         if tab is None:
-            # There is no active text editor.
+            # There is no active text editor. Exit.
             return
+        # Check the script's contents.
         python_script = tab.text().encode('utf-8')
         logger.debug('Python script:')
         logger.debug(python_script)
         # Check minification status.
         minify = False
         if uflash.get_minifier():
             minify = self.editor.minify
+        # Attempt and handle minification.
         if len(python_script) >= 8192:
```

Comment:
I believe we shouldn't change this limit *right now* since it's a known and expected limit. 

My concern is what should be a very simple process (getting MicroPython+scripts onto the device) is turning into a complicated ball of spaghetti code that we're just hacking about with. It makes me feel *very uncomfortable* to change so much behaviour in one go.

Let's just do the simplest possible solution to fulfil your requirements and then refactor.

Reply:
Okay.

Although whenever it is refactored, that values is not valid either as the script needs a 4 bytes header at the front, so it can only fit 8188 bytes of code.
Apart from that, assuming len() only returns the number of characters, it is possible to have utf-8 characters tha occupy more than 1 bye, so that could affect the calculation as well.

Reply:
Absolutely, but *now*, right before 1.0 release is NOT the time to try to fix this. It takes thought since it's more complicated than it at first appears and the current way things work is known (if slightly broken) behaviour.

FWIW... I'd rather get some feedback from micro:bit foundation on a blessed version of MicroPython *right now*.

I'm working my way through the outstanding issues in this PR and should be done mid-morning.

Reply:
> Absolutely, but now, right before 1.0 release is NOT the time to try to fix this. It takes thought since it's more complicated than it at first appears and the current way things work is known (if slightly broken) behaviour.

Of course, and I completely agree, that's why I said whenever it is refactored. I was just mentioning here as way to log it or communicate it. If you prefer I can open an issue instead.

> FWIW... I'd rather get some feedback from micro:bit foundation on a blessed version of MicroPython right now.

We are working right now on an updated version that correctly uses semver (thanks for pointing that out!). Pauline is on holiday and Jonny will send you an email as soon as he possibly can, as you made it perfectly clear who you expect that email from.

Reply:
A short "we're on it" email would've been helpful - its almost 10am and I'm none the wiser (hence my badgering you -- sorry). After all, the nub of my email is communication - even if that communication is "ack, don't worry we'll get back to you asap", that's better than nothing. I'm relieved this is being worked on... but to find out via the comments in a completely unrelated GitHub issue isn't exactly great communication is it..? (Sorry - I'm just finding this lack of communication VERY frustrating).